### PR TITLE
Modularize <iterator>

### PR DESCRIPTION
**Pull request by miscco**
This modularizes the `<iterator>` header as done by libc++.

This is a precursor to  updating to a more recent state of the code and bringing in C++20 ranges machinery

### CODE REVIEW

File: .upstream-tests/test/std/iterators/iterator.container/ssize.pass.cpp
Comment (griwes) [inline .upstream-tests/test/std/iterators/iterator.container/ssize.pass.cpp:147]: Why?

File: include/cuda/std/detail/libcxx/include/__iterator/reverse_iterator.h
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/__iterator/reverse_iterator.h:49]: Not really all that relevant to this PR, but since you've made me look through all this code: I wonder why this isn't changed with unstable ABI and if we should bite the bullet and remove it under unstable ABI in our copy.

File: .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp
Comment (griwes) [inline .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: Why? (I assume a compiler bug, but we should leave an explanation for this one here.)

**Review by griwes — APPROVED** [review]
Minor comments only, ship it.

File: .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp
Comment (miscco) [reply .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: Yeah there is a bug. That said it is so fundamental, that it shouldnt be a compiler bug.

In a nutshell with nvc++ the array s gets initialized to char s[] = "1111111111";

Why I have absolutely no idea. Given that it works on godbolt I assume something fishy with our test infra

File: .upstream-tests/test/std/iterators/iterator.container/ssize.pass.cpp
Comment (miscco) [reply .upstream-tests/test/std/iterators/iterator.container/ssize.pass.cpp:147]: This is again some very strange compiler bug that leads to wraparound on device

I should have added comments to those 🤦

File: include/cuda/std/detail/libcxx/include/__iterator/reverse_iterator.h
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/__iterator/reverse_iterator.h:49]: We should consider that in the future, but not in this PR

File: .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp
Comment (griwes) [reply .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: We should get to the bottom of this before merging this PR then.

Comment (miscco) [reply .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: That is preexisting, so I do not think it should prevent merging this

Comment (miscco) [reply .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: To be clear, in the current test config the following fails on device only
```cpp
int main(int, char**)
{
    char s[] = "1234567890";
    assert(s[5] == 6);
    return 0;
}
```

I do not know whether i configured it wrong, but nvc++ does some strange things

Comment (griwes) [reply .upstream-tests/test/std/iterators/predef.iterators/move.iterators/move.iter.ops/move.iter.op.index/difference_type.pass.cpp:50]: Okay, I didn't realize this was preexisting. We should figure this out fully, but I'm ok not blocking the PR over it.

**Pull request merged**
